lib/tls/session.toit

```diff
@@ -328,6 +471,132 @@ class Session:
+  reader_ /reader.BufferedReader
+  parent_ /Session
+
+  buffered_plaintext_index_ := 0
+  buffered_plaintext_ := []
+
+  constructor .parent_ .writer_ .reader_ .write_keys .read_keys:
+
+  write data from/int to/int -> int:
+    if to - from  == 0: return 0
+    // We want to be nice to the receiver in case it is an embedded device, so we
+    // don't send too large records.  This size is intended to fit in two MTUs on
+    // Ethernet.
+    List.chunk_up from to 2800: | from2/int to2/int length2 |
+      record_header := #[APPLICATION_DATA_, 3, 3, 0, 0]
+      BIG_ENDIAN.put_uint16 record_header 3 length2
+      // AES-GCM:
+      // The explicit (transmitted) part of the IV (nonce) must not be reused,
+      // but apart from that there are no requirements.  We just reuse the
+      // sequence number, which is a common solution.
+      // ChaCha20-Poly1305:
+      // There is no explict (transmitted) part of the IV, so it is a
+      // requirement that we use the sequence number to keep the local and
+      // remote IVs in sync.  As described in RFC8446 section 5.3 the serial
+      // number is xored with the last 8 bytes of the 12 byte nonce.
+      iv := write_keys.iv.copy
+      sequence_number /ByteArray := write_keys.next_sequence_number
+      explicit_iv /ByteArray := ?
+      if write_keys.has_explicit_iv:
+        explicit_iv = sequence_number
+        iv.replace 4 explicit_iv
+      else:
+        explicit_iv = #[]
+        8.repeat: iv[4 + it] ^= sequence_number[it]
+      encryptor := write_keys.new_encryptor iv
+      encryptor.start --authenticated_data=(sequence_number + record_header)
+      // Now that we have used the actual size of the plaintext as the authentication data
+      // we update the header with the real size on the wire, which includes some more data.
+      real_length := length2 + explicit_iv.size + Aead_.TAG_SIZE
+      BIG_ENDIAN.put_uint16 record_header 3 real_length
+      List.chunk_up from2 to2 512: | from3 to3 length3 |
+        first /bool := from3 == from2
+        last /bool := to3 == to2
+        plaintext := data is string
+            ? data.to_byte_array from3 to3
+            : data.copy from3 to3
+        parts := [encryptor.add plaintext]
+        if first:
+          parts = [record_header, explicit_iv, parts[0]]
+        if last:
+          parts.add encryptor.finish
+        else:
+          yield  // Don't monopolize the CPU with long crypto operations.
+        encrypted := byte_array_join_ parts
+        written := 0
+        while written < encrypted.size:
+          written += writer_.write encrypted written
+    return to - from
+
+  read -> ByteArray?:
+    result := null
+    try:
+      result = read_
+      return result
+    finally: | is_exception exception |
+      if not result or is_exception:
+        // If anything goes wrong we close the connection - don't want to give
+        // anyone a second chance to probe us with dodgy data.
+        parent_.close
```

Comment:
Is checking result really necessary in here? You're not passing a block to `read_`, so finally is called when you return result or get an exception. I think I'd just do:

```suggestion
    try:
      return read_
    finally: | is_exception exception |
      if is_exception:
        // If anything goes wrong we close the connection - don't want to give
        // anyone a second chance to probe us with dodgy data.
        parent_.close
```